### PR TITLE
[SPARK-23584][SQL] NewInstance should support interpreted execution

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr supported interpreted mode for `NewInstance`.

## How was this patch tested?
Added tests in `ObjectExpressionsSuite`.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala [hunk 1]
```diff
@@ -138,4 +154,40 @@ class ObjectExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
       checkEvaluation(decodeUsingSerializer, null, InternalRow.fromSeq(Seq(null)))
     }
   }
+
+  // This is an alternative version of `checkEvaluation` to compare results
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala]: This function was copy from #20757, so I'll remove later.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 2]
```diff
@@ -392,8 +392,44 @@ case class NewInstance(
...
 
-  override def eval(input: InternalRow): Any =
-    throw new UnsupportedOperationException("Only code-generated evaluation is supported.")
+  private lazy val constructor: (Seq[AnyRef]) => Any = {
+    val paramTypes = arguments.map { expr =>
+      CallMethodViaReflection.typeMapping.getOrElse(expr.dataType,
+        Seq(expr.dataType.asInstanceOf[ObjectType].cls))
+    }
+    val findConstructor = (types: Seq[Seq[Class[_]]]) => {
+      val constructorOption = cls.getConstructors.find { c =>
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Can we directly call `cls.getConstructor` with required `Class` to get it?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: ok, I'll re-check.

Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: How does this compare to JVM resolution rules? The code generated version will follow those.

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Review by hvanhovell — APPROVED** [review]
LGTM - merging to master. Thanks!

**Pull request closed without merging**